### PR TITLE
References resolution for UA counter-styles

### PR DESCRIPTION
**Pull request by vitorroriz**
#### 853c4e577730e9028620990bd328d13713b6f8ed
<pre>
References resolution for UA counter-styles
<a href="https://bugs.webkit.org/show_bug.cgi?id=252899">https://bugs.webkit.org/show_bug.cgi?id=252899</a>
rdar://103021161

Reviewed by Tim Nguyen.

We need to resolve references for the predefined counter-styles
which rules are stored in UA stylesheet.
These references can be resolved after the counter-style UA stylesheet
is parsed and before resolving references of the author counter styles.

* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::translateSuffixFromStyleProperties):
(WebCore::extractDataFromSystemDescriptor):
- Small adjustments in the default values for the descriptors.

(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem const):
(WebCore::CSSCounterStyleDescriptors::isValid const):
- Adding functions for validating a descriptor. For now, we only validate
the symbols according to the counter-style system, as described in the spec.

* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::resolveUserAgentReferences):
(WebCore::CSSCounterStyleRegistry::resolveReferencesIfNeeded):
(WebCore::CSSCounterStyleRegistry::resolveExtendsReference):
(WebCore::CSSCounterStyleRegistry::resolveFallbackReference):
(WebCore::CSSCounterStyleRegistry::counterStyle):
(WebCore::CSSCounterStyleRegistry::resolvedCounterStyle):
- Since the counter-styles map (userAgentCounterStyles()) lives in static memory,
I&apos;m modifying the resolution functions from the CSSCounterStyleRegistry for
being static and receiving a map instead, so they can serve for both
the UA and author counter styles. CSSCounterStyleRegistry::resolveUserAgentReferences()
is the wrapper for when we are resolving references for the UA counter-styles.

* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::StyleRuleCounterStyle::create):
* Source/WebCore/css/CSSCounterStyleRule.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeCounterStyleRule):
- Calling the new validation function for the descriptor before trying
to cerate a StyleRuleCounterStyles.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleNameInPrelude):
- The spec says that certain names, reserved for predefined counter-styles,
should not be accepted as a valid name during parser and if matched should
invalidate the rule. This is already done, but since we want to re-use
the same parsing functions for both author and UA rules we need a small tweak:
We are now passing the CSSParserMode to consumeCounterStyleNameInPrelude()
and allowing these reserved names when we are in UA mode.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::addToCounterStyleRegistry):
- Resolving UA counter-styles references after all counter-styles
are collected from the UA stylesheet to the UA counter-styles map.

* LayoutTests/webexposed/counter-style-image-symbols-not-exposed-expected.txt:
- Rebaselining. Although we parse images as symbols, we don&apos;t support it in counter-style
so the descriptor is invalidated.

Canonical link: <a href="https://commits.webkit.org/260940@main">https://commits.webkit.org/260940@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3b8355e3559293b3a6b747669162a7bd273051d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1455 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119052 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113965 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20578 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10305 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102270 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115758 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43543 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11815 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31530 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12437 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17805 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51142 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7584 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14232 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 1]
```diff
@@ -4714,7 +4714,7 @@ AtomString consumeCounterStyleNameInPrelude(CSSParserTokenRange& prelude)
     // In the context of the prelude of an @counter-style rule, a <counter-style-name> must not be an ASCII
     // case-insensitive match for "decimal" or "disc". No <counter-style-name>, prelude or not, may be an ASCII
     // case-insensitive match for "none".
-    if (identMatches<CSSValueDecimal, CSSValueDisc, CSSValueNone>(nameToken.id()))
+    if (mode != CSSParserMode::UASheetMode && identMatches<CSSValueDecimal, CSSValueDisc, CSSValueNone>(nameToken.id()))
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: Can we separate the case for `none` for correctness?

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp [hunk 2]
```diff
@@ -4777,7 +4777,8 @@ AtomString consumeCounterStyleNameInPrelude(CSSParserTokenRange& prelude)
     // In the context of the prelude of an @counter-style rule, a <counter-style-name> must not be an ASCII
     // case-insensitive match for "decimal" or "disc". No <counter-style-name>, prelude or not, may be an ASCII
     // case-insensitive match for "none".
-    if (identMatches<CSSValueDecimal, CSSValueDisc, CSSValueNone>(nameToken.id()))
+    auto id = nameToken.id();
+    if (id == CSSValueNone || (mode != CSSParserMode::UASheetMode && identMatches<CSSValueDecimal, CSSValueDisc, CSSValueNone>(id)))
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp]: `if (id == CSSValueNone || (mode != CSSParserMode::UASheetMode && identMatches<CSSValueDecimal, CSSValueDisc>(id)))`

File: Source/WebCore/css/CSSCounterStyleDescriptors.cpp [hunk 3]
```diff
@@ -229,4 +229,29 @@ CSSCounterStyleDescriptors CSSCounterStyleDescriptors::create(AtomString name, c
     return descriptors;
 }
 
+bool CSSCounterStyleDescriptors::areSymbolsValidForSystem() const
```
Comment (nt1m) [inline Source/WebCore/css/CSSCounterStyleDescriptors.cpp:232]: Changes in this file look unrelated

File: Source/WebCore/css/CSSCounterStyleDescriptors.h [hunk 4]
```diff
@@ -59,7 +59,7 @@ struct CSSCounterStyleDescriptors {
     };
     struct Pad {
         unsigned m_padMinimumLength = 0;
-        Symbol m_padSymbol = "-"_s;
+        Symbol m_padSymbol;
```
Comment (nt1m) [inline Source/WebCore/css/CSSCounterStyleDescriptors.h:62]: ditto for this file

File: Source/WebCore/css/CSSCounterStyleRegistry.cpp [hunk 5]
```diff
@@ -114,30 +126,33 @@ RefPtr<CSSCounterStyle> CSSCounterStyleRegistry::decimalCounter()
...
-
-    return decimalCounter();
+    // If there is a map, the search starts from the given map.
+    if (map) {
+        auto counter = getCounter(name, *map);
+        if (counter)
+            return counter;
+    }
+    // If there was no map (called for user-agent references resolution), or the counter was not found in the given map, we search at the user-agent map.
+    auto& userAgentMap = userAgentCounterStyles();
```
Comment (nt1m) [inline Source/WebCore/css/CSSCounterStyleRegistry.cpp]: You probably don't need the intermediary variable here, you can use it directly in `getCounter`

File: Source/WebCore/css/CSSCounterStyleDescriptors.cpp (hunk 3 above)
Comment (vitorroriz) [reply Source/WebCore/css/CSSCounterStyleDescriptors.cpp:232]: we are using this to invalidate a descriptor, I've clarified in the commit message, it is a small fix I've noticed when testing this patch

File: Source/WebCore/css/CSSCounterStyleDescriptors.h (hunk 4 above)
Comment (vitorroriz) [reply Source/WebCore/css/CSSCounterStyleDescriptors.h:62]: Since I was here I made a tiny initialization adjustment.
